builtin/provisioners/puppet/resource_provisioner.go

```diff
@@ -259,16 +259,29 @@ func (p *provisioner) generateAutosignToken(certname string) (string, error) {
 }
 
 func (p *provisioner) installPuppetAgentOpenSource() error {
+	task := "puppet_agent::install"
+
+	connType := p.instanceState.Ephemeral.ConnInfo["type"]
+	if connType == "" {
+		connType = "ssh"
+	}
+
+	agentConnInfo := map[string]string{
+		"type": connType,
+		"host": p.instanceState.Ephemeral.ConnInfo["host"],
+		"user": p.instanceState.Ephemeral.ConnInfo["user"],
+	}
+
 	result, err := bolt.Task(
-		p.instanceState.Ephemeral.ConnInfo,
+		agentConnInfo,
```

Comment:
🤔 I'm very sorry that I did not realize this at the same time as my previous comment, but it appears that there is other connection info used by `bolt.Task`:
https://github.com/hashicorp/terraform/blob/cf16b421356916a661aadc2af26c64f91fe7be03/builtin/provisioners/puppet/bolt/bolt.go#L46

I think we should continue to send the entire `p.instanceState.Ephemeral.ConnInfo` to `bolt.Task`, and move this logic into  `bolt.Task`. What do you think?

Reply:
@mildwonkey no don't be sorry, I should have noticed it earlier. I am in two minds- there are 2 calls to bolt.Task. In the first, the connInfo is built up iteratively making everything explicit. In the second, connInfo is being passed as-is making bolt.Task a bit of a black box. That's what confused me just now too. Meanwhile once you're inside bolt.Task, it is not clear to the reader there either that sometimes they are receiving an `p.instanceState.Ephemeral.ConnInfo` object and other times some data that was composed statically. So, if I define everything that bolt.Task needs in agentConnInfo I think it will be clearer. Of course, your suggestion could be more maintainable- I'm not certain. What do you think ?

Reply:
🤷‍♀You make an excellent point, I like your suggestion. Both approaches are valid, and as I don't have any insight into future plans for this provisioner (if there are any), I think you can go with whatever makes the most sense to you. 

And honestly, this is a fairly easy-to-refactor detail so I don't think we'll go wrong either way!